### ISSUE TITLE
Question about the gradient of QTensor and QBitTensor

### ISSUE THREAD

**Issue by shuokay** [huggingface/quanto#146]
I am confused by the gradient of the Quantizer and QBitTensor. Take QTensor as the example:

The evaluation of forward is:
```txt
data  = base / scale  (1)
data = round(data)  (2)
data = clamp(data, qmin,  qmax)  (3)
```
I think the graidents should be:
```txt
grad_div = 1 / scale  (1)
grad_round = 1  (2)  # refer to "straight though estimator": https://arxiv.org/abs/1308.3432
grad_clamp = 1 if qmin < data < qmax else 0  (3)
```
According to chain rule, the gradient of Quantizer should be `grad_div * grad_round * grad_clamp` which is equal to `1 / scale if qmin < base/scale < qmax else 0`

I have reached QTensor's unit test and I find that dequantize is applied to QTensor before backward. I am confused by `Quantizer. backward` and the  `dequantize` behavior before backward.

**Comment by dacorvo** [comment]
The first thing to understand is that this applies only to QAT, where the original float weights are kept and quantized on-the-fly in the forward graph using a static scale and zeropoint.
The gradient of the whole quantize operation is indeed a Straight-Through Estimator (STE), meaning that the underlying float weight will receive the same gradient as they would have received if they were not quantized.
What you are describing is a finer grained STE that would result in evaluating gradients for the scale and zeropoint, which is in my experience actually counter-productive due to adversarial evolutions of the underlying weights and the quantization boundaries.

**Comment by shuokay** [comment]
@dacorvo Thanks for your reply. I can understand most of your explanation, but can you further explain "adversarial evolutions of the underlying weights and the quantization boundaries."

**Comment by dacorvo** [comment]
This is basically like the "sawing the branch you are sitting on". The weights gradients will try to move them to the upper or lower quantization bound to minimize loss, but in the meantime the scale and zeropoint will be modified to move these boundaries.

**Comment by shuokay** [comment]
> This is basically like "sawing the branch you are sitting on". The weights gradients will try to move them to the upper or lower quantization bound to minimize loss, but in the meantime the scale and zeropoint will be modified and these boundaries will move.

Hi @dacorvo, I appreciate your patient response to my question.
About quantization boundary, I think the boundary problem is caused by the `clamp` operator and I have had tried to change  the gradient from `1  if qmin <= quantized_value <= qmax else 0` to always be `1` on some models years ago, however  there is no obvious winner between them, which maybe conflict with your experience.

About the STE, I reread the paper, and it said:
>The idea is simply to back-propagate through the hard threshold function (1 if the argument is positive, 0 otherwise) as if it had been the identity function.

So I think the STE should only be applied to threshold functions whose gradient is zero and lead the backpropagation learning failure.  I still think that the gradient of Quantize should be `1 / scale if qmin < base/scale < qmax else 0`. 
I have run the `quantize_minist_model.py` example and it works well indeed, however, I suspect that this model may be too simple to expose the problem. I plan to conduct a slightly more complex experiment, such as using MobileNetV2 with the CIFAR-10 dataset

**Comment by dacorvo** [comment]
Trust me, there is no problem. I already fine-tuned many models like that, including ViT

**Comment by shuokay** [comment]
Regarding the gradient of the `Quantizer` in Quanto being `1` instead of `1/scale` (similarly for the `Dequantizer`), I conducted some small experiments, searched through the PyTorch source code, and added some print statements in Quanto. Based on these investigations, I can confidently affirm that Quanto's approach is correct. To address any potential confusion among others, I'd like to share my understanding here:

1. It's essential to comprehend the gradients of the Quantizer/Dequantizer in Quanto within the context of the entire framework rather than examining them in isolation. While focusing solely on the Quantizer's gradient may lead to confusion, the overall framework is indeed correct.

2. The crucial aspect lies in Quanto's fallback dispatcher (named `qfallback`), which ensures that Quanto always evaluates the gradient in the float domain. This approach is driven by several factors:
   - When computing the loss, Quanto requires dequantization of the model's output, resulting in the loss being in the float domain.
   - Torch's `grad_fn` typically adopts the naming convention `xxx_backward`. However, Quanto never registers the dispatcher entry for `xxx_backward`. Consequently, the calculation of backward always falls back to the float domain, just as if there's no `scale` applied.
   - According to the Straight-Through Estimator (STE) theory, the gradients of the Quantizer and Dequantizer are expected to be `1`.

cc @dacorvo
